File: src/main/java/raylras/zen/util/DebugUtils.java
package raylras.zen.util;

import org.antlr.v4.runtime.RuleContext;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;
import raylras.zen.code.parser.ZenScriptLexer;
import raylras.zen.code.parser.ZenScriptParser;

public class DebugUtils {

    public static String prettyPrintTree(ParseTree node) {
        StringBuilder builder = new StringBuilder();
        prettyPrintTree(builder, "", node, true);
        return builder.toString();
    }

    private static void prettyPrintTree(StringBuilder builder, String indent, ParseTree node, boolean last) {
        if (node == null) {
            return;
        }

        String text;
        if (node instanceof TerminalNode) {
            Token token = ((TerminalNode) node).getSymbol();
            int tokenType = token.getType();
            if (tokenType == -1) {
                text = "<EOF>";
            } else {
                text = ZenScriptLexer.ruleNames[tokenType - 1] + ":" + token.getText();
            }
        } else if (node instanceof RuleContext){
            int ruleIndex = ((RuleContext) node).getRuleIndex();
            text = ZenScriptParser.ruleNames[ruleIndex];
        } else {
            text = null;
        }

        builder.append(indent).append(last ? "\\-- " : "|-- ").append(text).append("\n");

        int childCount = node.getChildCount();
        for (int i = 0; i < childCount; i++) {
            prettyPrintTree(builder, indent + (last ? "    " : "|   "), node.getChild(i), i == childCount - 1);
        }
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/util/DebugUtils.kt
package raylras.zen.util

import org.antlr.v4.runtime.RuleContext
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.TerminalNode
import raylras.zen.model.parser.ZenScriptLexer
import raylras.zen.model.parser.ZenScriptParser

object DebugUtils {
    fun prettyPrintTree(node: ParseTree?): String {
        val builder = StringBuilder()
        prettyPrintTree(builder, "", node, true)
        return builder.toString()
    }

    private fun prettyPrintTree(builder: StringBuilder, indent: String, node: ParseTree?, last: Boolean) {
        if (node == null) {
            return
        }

        val text: String?
        if (node is TerminalNode) {
            val token = node.symbol
            val tokenType = token.type
            text = if (tokenType == -1) {
                "<EOF>"
            } else {
                ZenScriptLexer.ruleNames[tokenType - 1] + ":" + token.text
            }
        } else if (node is RuleContext) {
            val ruleIndex = node.ruleIndex
            text = ZenScriptParser.ruleNames[ruleIndex]
        } else {
            text = null
        }

        builder.append(indent).append(if (last) "\\-- " else "|-- ").append(text).append("\n")

        val childCount = node.childCount
        for (i in 0 until childCount) {
            prettyPrintTree(builder, indent + (if (last) "    " else "|   "), node.getChild(i), i == childCount - 1)
        }
    }
}
